src/libraries/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs

```diff
@@ -246,59 +250,68 @@ internal bool Contains(HeaderDescriptor descriptor)
 
         public override string ToString()
         {
-            if (_headerStore == null || _headerStore.Count == 0)
-            {
-                return string.Empty;
-            }
-
             // Return all headers as string similar to:
             // HeaderName1: Value1, Value2
             // HeaderName2: Value1
             // ...
-            var sb = new StringBuilder();
-            foreach (KeyValuePair<string, string> header in GetHeaderStrings())
-            {
-                sb.Append(header.Key).Append(": ").AppendLine(header.Value);
-            }
 
-            return sb.ToString();
-        }
+            var vsb = new ValueStringBuilder(stackalloc char[512]);
 
-        internal IEnumerable<KeyValuePair<string, string>> GetHeaderStrings()
-        {
-            if (_headerStore == null)
+            if (_headerStore is Dictionary<HeaderDescriptor, object> headerStore)
             {
-                yield break;
-            }
+                foreach (KeyValuePair<HeaderDescriptor, object> header in headerStore)
+                {
+                    vsb.Append(header.Key.Name);
+                    vsb.Append(": ");
 
-            foreach (KeyValuePair<HeaderDescriptor, object> header in _headerStore)
-            {
-                string stringValue = GetHeaderString(header.Key, header.Value);
+                    GetStoreValuesAsStringOrStringArray(header.Key, header.Value, out string? singleValue, out string[]? multiValue);
+                    Debug.Assert(singleValue is not null ^ multiValue is not null);
+
+                    if (singleValue is not null)
+                    {
+                        vsb.Append(singleValue);
+                    }
+                    else
+                    {
+                        // Note that if we get multiple values for a header that doesn't support multiple values, we'll
+                        // just separate the values using a comma (default separator).
+                        HttpHeaderParser? parser = header.Key.Parser;
+                        string? separator = HttpHeaderParser.DefaultSeparator;
+                        if (parser != null && parser.SupportsMultipleValues)
+                        {
+                            separator = parser.Separator;
+                        }
```

Comment:
Nit:
The other 2 places choosing a separator in this PR are written using a ternary
```suggestion
                        string? separator = header.Key.Parser is HttpHeaderParser parser && parser.SupportsMultipleValues ? parser.Separator : HttpHeaderParser.DefaultSeparator;
```